test/apps/iodemo/io_demo.cc

```diff
@@ -1309,8 +1310,12 @@ class DemoServer : public P2pDemoCommon {
         w->init(this, conn, msg->sn, msg->conn_id, iov,
                 &conn_stat.completions<IO_WRITE>());
         conn_stat.bytes<IO_WRITE>() += msg->data_size;
-        conn->recv_am_data((*iov)[0].buffer(), (*iov)[0].size(),
-                           (*iov)[0].memh(), data_desc, w);
+        if (!ucx_am_is_rndv(data_desc)) {
+            conn->recv_am_data(NULL, 0, NULL, data_desc, w);
+        } else {
```

Comment:
why needed?
IOV must be valid after `prepare_am_recv_data_iov()`: https://github.com/openucx/ucx/blob/325a608ad53423a758f6e6944feb067cecefe1b3/test/apps/iodemo/io_demo.cc#L1023

Reply:
in !ucx_am_is_rndv(data_desc) case, _iov.size() would be 0 but the assertion you pointed out still stands because of (!!_extra_buf == 1): https://github.com/openucx/ucx/blob/325a608ad53423a758f6e6944feb067cecefe1b3/test/apps/iodemo/io_demo.cc#L614

Reply:
I see, that makes code weird.
- `iov::size()` considers both `extra_buf` and `iov`
- `iov::operator[]()` considers only `iov`

I'd either remove `extra_buf` and save AM desc to `iov[0]`.
or update operator[] as follow (to aligh `operator[]()` and `size()` methods of `BufferIov` class):
```
inline Buffer &operator[](size_t i) const
{
    if (!_iov.empty()) {
        return *_iov[i];
    } else {
        assert(_extra_buf != NULL);
        return _extra_buf;
    }
}
```

Reply:
in addition, all other `BufferIov` works fine with both `extra_buf` and `iov`, only `operator[]()` _stands out of the crowd_

Reply:
_extra_buf can not be returned directly because it is not Buffer.
_extra_buf is hard to be wrapped in Buffer because _extra_buf is allocated and freed by UCX instead of io_demo.
we could use a bogus Buffer to represent _extra_buf in this case if that works for you. (a new commit submitted for this idea)

Reply:
> we could use a bogus Buffer to represent _extra_buf in this case if that works for you. (a new commit submitted for this idea)

IMO, good idea